keps/sig-node/3619-supplemental-groups-policy/README.md

```diff
@@ -0,0 +1,857 @@
+  containers:
+  - image: some-image # uid=1000 belongs to gid=60000 in the image
+    sh: ["id"]
+```
+Then, groups of the first process of the container will be :
+- `1000`: specified by `runAsGroup`
+- `50000`: specified by `supplementalGroups`
+- `60000`: defined in the image (<strong><em>Note: Kubernetes keep this</em></strong>)
+
+Thus, the container logs will be:
+
+```console
+uid=1000(...) gid=1000(...) groups=1000(...),50000,60000(...)
+```
+
+This KEP proposes a complementary field named `SupplementalGroupsPolicy` that enables users to customize to respect/ignore the group membership of the primary UID of the container defined in the container image when applying `SupplementalGroups` to the container processes. 
+
+## Motivation
+
+As described above, `SupplementalGroups` will be just added to the groups defined in the container image for the container's primary UID. When a cluster enforces some security policy for pods that protects the value of `SupplementalGroups`, the effect of its enforcement is limited, i.e., cluster users can easily bypass the policy enforcement just by using a custom image. If such a bypass happened, it would be unexpected behavior for most cluster administrators because the enforcement is almost useless. Moreover, the bypass will cause unexpected file access permission. In some use cases, the unexpected file access permission will be a security concern. For example, using `hostPath` volumes could be a severe problem because UID/GIDs matter in accessing files/directories in the volumes.
+
+However, Kubernetes provides no API surface to prevent this bypass, although it could sometimes lead to a security concern. Because the behavior is implemented in CRI implementations actually, To mitigate this, the cluster administrators will need to deploy a custom low-level container runtime(e.g., [pfnet-research/strict-supplementalgroups-container-runtime](https://github.com/pfnet-research/strict-supplementalgroups-container-runtime)) that modifies OCI container runtime spec(`config.json`) produced by CRI implementations (e.g., containerd, cri-o). A custom `RuntimeClass` would be introduced for it. Nevertheless, It would be an extra operational burden for cluster administrators.
+
+Thus, this KEP proposes to offer a new API field named `SupplementalGroupsPolicy` that customize to respect/ignore the group membership of the primary UID of the container defined in the container image when applying `SupplementalGroups` to the container processes. The new API allows cluster administrators to deploy security policies that protect the `SupplementalGroupsPolicy` field in the cluster to avoid the unexpected bypass of `SupplementalGroups` described above.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+To Provide a new API field to control exactly which groups the container process belongs to.
+
+### Non-Goals
+
+<!--
```

Comment:
Is it a goal to not change OCI spec? Or we want it in goals to make sure that OCI spec is clear(-er)?

```suggestion
 - Change the OCI specification
 
<!--
```

Reply:
As tim verified in [the thread](https://github.com/kubernetes/enhancements/pull/3620#discussion_r1089848576), containerd is ok via containerd's API directly(by nerdctl), however containerd via cri  (by kubernetes) is not ok.

So, I think OCI spec(both on container and image) is clear enough:
- [`User in config.json`](https://github.com/opencontainers/runtime-spec/blob/main/config.md#user), and 
- [`config.User` in image spec](https://github.com/opencontainers/image-spec/blob/3a7f492d3f1bcada656a7d8c08f3f9bbd05e7406/config.md#:~:text=User%20string%2C%20OPTIONAL).

In my understanding, we should fix(make clearer) k8s API and CRI specs:

- [PodSecurityContext.RunAsGroup](https://github.com/everpeace/kubernetes/blob/master/pkg/apis/core/types.go#L3116-L3123)
- [PodSecurityContext.SupplementalGroups](https://github.com/everpeace/kubernetes/blob/master/pkg/apis/core/types.go#L3133-L3138)
- [SecurityContext.RunAsGroup](https://github.com/everpeace/kubernetes/blob/master/pkg/apis/core/types.go#L5487-L5493)
- [LinuxSandboxSecurityContext.run\_as\_group](https://github.com/kubernetes/cri-api/blob/master/pkg/apis/runtime/v1/api.proto#L321-L323)
- [LinuxSandboxSecurityContext.supplemental\_groups](https://github.com/kubernetes/cri-api/blob/master/pkg/apis/runtime/v1/api.proto#L326-L332)
- [LinuxContainerSecurityContext.run\_as\_group](https://github.com/kubernetes/cri-api/blob/master/pkg/apis/runtime/v1/api.proto#L869-L872)
- [LinuxContainerSecurityContext.supplemental\_groups](https://github.com/kubernetes/cri-api/blob/master/pkg/apis/runtime/v1/api.proto#L879-L885).

Reply:
oh, I misunderstood a bit. Still catching up. I thought there was comment that OCI spec wasn't 100% precise.

Reply:
oci spec does not cover k8s, cri api, or docker cli options.. 

oci specs are .. lower level

Reply:
OCI image-spec (https://github.com/opencontainers/image-spec/blob/main/config.md)  says:

"""
If group/gid is not specified, the default group and supplementary groups of the given user/uid in /etc/passwd from the container are applied.
"""

I think it should be updated to something like:

"""
If group/gid is not specified, the default group and supplementary groups of the given user/uid in /etc/passwd and /etc/group from the container are applied.  If group/gid is specified, supplementary groups from the container are ignored.
"""

Additionally, https://github.com/opencontainers/image-spec/blob/main/conversion.md says:

"""
a converter SHOULD set the value of process.user.additionalGids to a value corresponding to the user in the container's context described by Config.User. For Unix-like systems, this MAY involve resolution through NSS or parsing /etc/group and determining the group memberships of the user specified in process.user.uid. If the value of user in Config.User is numeric, the converter SHOULD NOT modify process.user.additionalGids
"""

I think this needs to be updated to something like:

"""
a converter SHOULD set the value of process.user.additionalGids to a value corresponding to the user in the container's context described by Config.User. For Unix-like systems, this MAY involve resolution through NSS or parsing /etc/group and determining the group memberships of the user specified in process.user.uid. If the value of user in Config.User is numeric or if Config.user specifies a group, the converter SHOULD NOT modify process.user.additionalGids
"""

That clarifies the existing behaviors.  The question remains as to whether OCI should say anything about `additionalGids` when `Config.user` does not specify a group. 

IOW: `docker` and `nerdctl` seem to define `--group-add` as "adding a group on top of the default OCI behavior".  I think Kubernetes should define `supplementalGroups` as "replacing the OCI default behavior", whether `runAsGroup` is defined or not.